Detect shared library name with llvm-config

RyanGlScott opened the pull request:
This is the same pull request as https://github.com/bscarlet/llvm-general/pull/203, but since all the latest development seems to be happening here, I thought I'd make a PR here as well. I'll recap the issue below:

Currently, `llvm-hs`'s `Setup.hs` script hardcodes the name of of the shared LLVM library based on its version number (e.g., with `llvm-3.9`, it hardcodes `LLVM-3.9`). The problem is that the MinGW-w64 (Windows) version of LLVM doesn't put version numbers at the end of its library files - that is, it names it `libLLVM.dll.a` instead of `libLLVM-3.9.dll.a`. As a result, `llvm-hs` fails to configure on Windows because it can't find the right shared library name:

```
$ cabal configure -fshared-llvm
Resolving dependencies...
Configuring llvm-general-3.9.0.0...
Warning: Instead of 'cc-options: -IC:\msys64\mingw64/include' use
'include-dirs: C:\msys64\mingw64/include'
setup.exe: Missing dependency on a foreign library:
* Missing C library: LLVM-3.9.1
This problem can usually be solved by installing the system package that
provides this library (you may need the "-dev" version). If the library is
already installed but in a non-standard location then you can use the flags
--extra-include-dirs= and --extra-lib-dirs= to specify where it is.
```

See https://github.com/AccelerateHS/accelerate/pull/342#issuecomment-274116175 for an example of where this problem arose (with `llvm-general`).

What seems like the obvious solution here is to instead just determine what the shared library name is from `llvm-config`, like `Setup.hs` currently does for the static library names. This PR does just that.

Note: Since I've only been able to successfully install a shared build of LLVM on Windows using version 3.9, I opened this PR on the `llvm-3` branch. If you'd prefer that it be done differently, please let me know.

RyanGlScott commented:
I'll confess that I don't quite understand why the Travis build failed, especially since I verified that this builds for me on a Windows and Linux machine.

RyanGlScott committed:
Detect shared library name with llvm-config

cocreature commented:
This looks great, thanks! The travis failure definitely doesn’t look like your fault, I’m attempting a fix in #24. I’ll merge once I manage to fix travis.

> Note: Since I've only been able to successfully install a shared build of LLVM on Windows using version 3.9, I opened this PR on the llvm-3 branch. If you'd prefer that it be done differently, please let me know.

That’s fine, I can apply it to the `llvm-4` branch after we merge this. The changes don’t look specific However, you should be aware that we currently do not have plans for a hackage release of the `llvm-3` branch. If that’s important to you, we can reconsider this.

FYI, I’m currently busy with exams, so I’m sorry if it takes a few days until I’m able to respond.

cocreature commented:
This looks great, thanks! The travis failure definitely doesn’t look like your fault

> I'll confess that I don't quite understand why the Travis build failed, especially since I verified that this builds for me on a Windows and Linux machine.

I was wrong before, the Travis failure is related to this pull request. The problem is that `llvm-config --libs --link-shared` fails if the LLVM shared library is not available (and it is not on travis). Since LLVM doesn’t build shared libraries by default this is a configuration that we should support imho. We thus need to handle the error in that case. I haven’t yet thought about the best way to do so.

RyanGlScott commented:
> The problem is that `llvm-config --libs --link-shared` fails if the LLVM shared library is not available (and it is not on travis).

Ah, that makes sense now. Then I suppose we should only run `llvm-config --libs --link-shared` if `-fshared-llvm` is actually enabled. (I'm a bit surprised that `llvm-config` command actually gives an _error_ if you don't have a shared LLVM build, but whaddya gonna do, I suppose.)

I'll need to do some investigation to see what the best way of conditionally running an `IO` action based on the value of a `Cabal` flag in a `Setup.hs` script is.

RyanGlScott committed:
Only run llvm-config --libs --link-shared if -fsystem-llvm is enabled

As observed in https://travis-ci.org/llvm-hs/llvm-hs/builds/205219107, if you
run llvm-config --libs --link-shared with an LLVM installation that doesn't
have shared libraries, it will simply fail.

The workaround is to simply check whether the user enabled -fshared-llvm before
deciding between --link-shared and --link-static. That way, Setup.hs will only
error if the user explicitly asks to use a shared LLVM installation when they
don't have one (as opposed to before, where it would happen even if Setup.hs
ultimately decided on a static build).

RyanGlScott commented:
OK, that wasn't so bad. I now explicitly check the value of the `shared-llvm` flag beforehand to determine whether to pass `--link-shared` or `--link-static` to `llvm-config`. That way, if you use `--link-shared` and you don't have a shared LLVM installation, it's your own fault. :)

cocreature commented:
Thanks, this looks great! I like how this is actually simpler than before :)
